NAS-116414: ZFS Info Card

denysbutenko opened the pull request:
<img width="381" alt="image" src="https://user-images.githubusercontent.com/351613/177195844-9ec6faaa-56c0-4561-bf7f-3006d807ac55.png">

For testing go to Devices page and select an disk

bugclerk commented:
Jira URL: https://ixsystems.atlassian.net/browse/NAS-116414

RehanY147 commented on src/app/pages/storage2/modules/devices/components/zfs-info-card/zfs-info-card.component.ts:
```diff
@@ -0,0 +1,178 @@
+
+  ngOnInit(): void {
+    this.poolId = +this.route.snapshot.paramMap.get('poolId');
+    this.slideInService.onClose$?.pipe(untilDestroyed(this)).subscribe(() => {
```
This will likely always reload disk even if the model is closed by clicking cancel/cross button. Let's make sure all instances of `sliceInService.close()` in the form return `true` for successful changes and then check here for that boolean. So we can only reload when required and not always.

denysbutenko replied:
Sounds reasonable but it seems there are no places where it is implemented. I'll keep consistency with DiskInfoWidget and other forms for now.

RehanY147 commented on src/app/pages/storage2/modules/devices/components/zfs-info-card/zfs-info-card.component.ts:
```diff
@@ -0,0 +1,178 @@
+    ).subscribe(() => {
+      this.ws.call('pool.offline', [this.poolId, { label: this.topologyItem.guid }]).pipe(untilDestroyed(this)).subscribe(
+        () => {
+          this.devicesStore.reloadList();
```
These subscribe and error call backs are same in `onOnline`, `onOffline` and `onDetach`. Let's create named functions vars and use them as call backs to avoid redundancy.

denysbutenko replied:
Could you show an example?

denysbutenko replied:
It wont work, keep as is

RehanY147 reviewed (changes requested):
When I detach a disk from a mirror, the table doesn't reload. Also, the details pain continues to show old disk data that was removed. Also, when detach action is confirmed, there's no loading indication for a while. Nothing happens until the operation is complete. This might be a change detection issue.

denysbutenko commented:
> When I detach a disk from a mirror, the table doesn't reload. 
Added a call to reload data on success

> Also, the details pain continues to show old disk data that was removed. 

Seems this issue not related to ZFS Info Card

> Also, when detach action is confirmed, there's no loading indication for a while. Nothing happens until the operation is complete. This might be a change detection issue.

Added loading indicator for operations

RehanY147 commented:
> Seems this issue not related to ZFS Info Card
Since, the remove/detach actions were introduces by the ZFS Info card, I think it should be handled there. Let me know if you'd like to discuss this.

RehanY147 reviewed (changes requested):
You can add the following filter pipe to `this.slideInService.onClose$.pipe(...)` in both `DiskInfoComponent` and `ZfsInfoComponent` to avoid loading disks even when the slide in was closed without any changes.
`filter((value) => !!value.response && value.modalType === DiskFormComponent)`

denysbutenko commented:
> You can add the following filter pipe to `this.slideInService.onClose$.pipe(...)` in both `DiskInfoComponent` and `ZfsInfoComponent` to avoid loading disks even when the slide in was closed without any changes.
> `filter((value) => !!value.response && value.modalType === DiskFormComponent)`

Ok, added

denysbutenko commented:
> > Seems this issue not related to ZFS Info Card
> 
> Since, the remove/detach actions were introduces by the ZFS Info card, I think it should be handled there. Let me know if you'd like to discuss this.

Ok, fixed

RehanY147 reviewed (approved):
Looks good. But let's also capitalize the `mirror-0` for `Parent` attribute. That's also capitalized on the pool status page on old storage page.
